### ISSUE TITLE
[R-package] training continuation unit test is broken

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#2468]
One unit test in the R package is currently broken. Steps to reproduce on Mac

```r
export CXX=/usr/local/bin/g++-8 CC=/usr/local/bin/gcc-8
Rscript build_r.R
cd R-package/tests
Rscript testthat.R
```

This results in the following error at the ends of the logs

```
[LightGBM] [Info] Saving data to binary file /var/folders/xq/wktq4zdx4jd3qdpk34d28m940000gn/T//RtmpiY1DzV/lgb.Dataset_15553eef826
══ testthat results  ═══════════════════════════════════════════════════════════
[ OK: 36 | SKIPPED: 0 | WARNINGS: 0 | FAILED: 1 ]
1. Error: training continuation works (@test_basic.R#71)
```

I'm uncertain when the issue was first introduced (we do not run R tests in CI yet, #2335), but this behavior was observed at least as of https://github.com/microsoft/LightGBM/commit/cc7a1e27398d86e9befdf6eb08141359882f3d49.

**Comment by jameslamb** [comment]
Looked into this some more tonight. I still haven't gotten to the root cause.

This was failing with basically an out-of-index error. When you do training continuation, it seems the second

```r
# check this out:
bst1$record_evals[[data_name]][[eval_name]]$eval
bst2$record_evals[[data_name]][[eval_name]]$eval
```

things you'll notice:

* best iteration from `bst1` is iteration 3
* if you compare the errors, it looks like `bst2` starts from iteration 3 (notice the 4th iteration in `bst1` and first in `bst2` have identical error)
* `bst2$current_iter()` is 8, also consistent with starting from the best iteration
* I think the issue is here in `lgb.train()`:

    ```r
    if (is.character(init_model)) {
        predictor <- Predictor$new(init_model)
    ```

The `Predictor` created from the initial model should be the same as the booster state after 5 iterations, I think, but it is not.

So basically you can get into a situation where the Booster thinks it has seen 8 iterations, but only 5 iterations of results are stored.

* `length(bst$record_evals[["train"]][["binary_logloss"]]$eval)` is 10
* `length(bst1$record_evals[["train"]][["binary_logloss"]]$eval)` is 5
* `length(bst2$record_evals[["train"]][["binary_logloss"]]$eval)` is 5
